File: trabalho_do_semestre_python/src/main/conexao/ProdutoDAO.py
from tkinter import messagebox
from conexao.FabricaConexao import FabricaConexao  
from dominio.Produto import Produto  
import mysql.connector  

class ProdutoDAO:
    # Método para inserir um novo produto no banco de dados
    def create(self, p):
        con = FabricaConexao.getConection()  # Obtém a conexão com o banco de dados
        stmt = None

        try:
            stmt = con.cursor()
            # Define a query SQL para inserção de um novo produto
            sql = "INSERT INTO PRODUTOS (NOME, DESCRICAO, MARCA, CATEGORIA, PRECO, CUSTO) VALUES (%s, %s, %s, %s, %s, %s)"
            val = (p.nome, p.descricao, p.marca, p.categoria, p.preco, p.custo)
            stmt.execute(sql, val)  
            con.commit()  # Realiza o commit da transação no banco de dados
        except mysql.connector.Error as e:
            FabricaConexao.trataProblemaBancoDados(e)  # Trata erros específicos do banco de dados
        finally:
            FabricaConexao.closeConnectionWithStmt(con, stmt)  # Fecha a conexão com o banco de dados

    # Método para ler todos os produtos do banco de dados
    def read(self):
        produtos = []  # Lista para armazenar os produtos lidos

        con = FabricaConexao.getConection()  # Obtém a conexão com o banco de dados
        stmt = None
        rs = None

        try:
            stmt = con.cursor()
            stmt.execute("SELECT * FROM PRODUTOS")  # Executa a query SQL para selecionar todos os produtos
            rows = stmt.fetchall() 

            for row in rows:
                produto = Produto()  # Cria um objeto Produto para cada linha retornada
                produto.ID = row[0]
                produto.nome = row[1]
                produto.descricao = row[2]
                produto.marca = row[3]
                produto.categoria = row[4]
                produto.preco = row[5]
                produto.custo = row[6]
                produtos.append(produto)  # Adiciona o produto à lista de produtos lidos

        except mysql.connector.Error as e:
            FabricaConexao.trataProblemaBancoDados(e)  # Trata erros específicos do banco de dados
        finally:
            FabricaConexao.closeConnectionWithStmtAndRs(con, stmt, rs)  # Fecha conexão e recursos

        return produtos  # Retorna a lista de produtos lidos do banco de dados

    # Método para atualizar um produto no banco de dados
    def update(self, p):
        con = FabricaConexao.getConection()  # Obtém a conexão com o banco de dados
        stmt = None

        try:
            stmt = con.cursor()
            # Define a query SQL para atualizar os dados de um produto existente
            sql = "UPDATE PRODUTOS SET NOME = %s, DESCRICAO = %s, MARCA = %s, CATEGORIA = %s, PRECO = %s, CUSTO = %s WHERE PRODUTOS_ID = %s"
            val = (p.nome, p.descricao, p.marca, p.categoria, p.preco, p.custo, p.ID)
            stmt.execute(sql, val)  
            con.commit()  # Realiza o commit da transação no banco de dados
        except mysql.connector.Error as e:
            FabricaConexao.trataProblemaBancoDados(e)  # Trata erros específicos do banco de dados
        finally:
            FabricaConexao.closeConnectionWithStmt(con, stmt)  # Fecha a conexão com o banco de dados

    # Método para deletar um produto do banco de dados
    def delete(self, p):
        con = FabricaConexao.getConection()  # Obtém a conexão com o banco de dados
        stmt = None

        try:
            stmt = con.cursor()
            # Define a query SQL para deletar um produto do banco de dados
            sql = "DELETE FROM PRODUTOS WHERE PRODUTOS_ID = %s"
            val = (p.ID,)
            stmt.execute(sql, val)  # Executa a query SQL com o ID do produto a ser deletado
            con.commit()  # Realiza o commit da transação no banco de dados
        except mysql.connector.Error as e:
            FabricaConexao.trataProblemaBancoDados(e)  # Trata erros específicos do banco de dados
        finally:
            FabricaConexao.closeConnectionWithStmt(con, stmt)  # Fecha a conexão com o banco de dados

File: trabalho_do_semestre_java/src/main/java/conexaoSQL/DAO/ProdutoDAO.java
package conexaoSQL.DAO;

import java.sql.*;
import java.util.*;

import conexaoSQL.FabricaConexao;
import modelos.*;

/**
 * Classe ProdutoDAO responsável por gerenciar as operações de acesso a dados
 * do Produto no banco de dados MySQL.
 *
 * Esta classe fornece métodos para criar, ler, atualizar e deletar registros
 * da tabela PRODUTOS presente no banco de dados.
 */
public class ProdutoDAO {

    /**
     * Insere um novo produto na tabela PRODUTOS.
     *
     * @param p o objeto Produto a ser inserido.
     */
    public void create(Produto p){
        // cria uma conexao com o banco de dados
        Connection con = FabricaConexao.getConection();
        // cria uma conexao para editar o banco de dados
        PreparedStatement stmt = null;

        try {
            // atribui os atributos do produto a ser inserido
            stmt = con.prepareStatement("INSERT PRODUTOS (NOME, DESCRICAO, MARCA, CATEGORIA, PRECO, CUSTO) VALUES(?, ?, ?, ?, ?, ?)");
            stmt.setString(1, p.getNome());
            stmt.setString(2, p.getDescricao());
            stmt.setString(3, p.getMarca());
            stmt.setString(4, p.getCategoria());
            stmt.setDouble(5, p.getPreco());
            stmt.setDouble(6, p.getCusto());
            // sexeuta
            stmt.executeUpdate();
        } catch (SQLException e) {
            FabricaConexao.trataProblemaBancoDados(e);
        } finally {
            //fecha a conexao criada
            FabricaConexao.closeConnection(con, stmt);
        }
    }

    /**
     * Le toda a tabela do banco de dados
     *
     * @param produtos a lista com os produtos á ser preenchida.
     * @return a lista de produtos preenchida com os registros da tabela PRODUTOS.
     */
    public List<Produto> read (List<Produto> produtos){

        // cria conexoes necesarias
        Connection con = FabricaConexao.getConection();
        PreparedStatement stmt = null;
        ResultSet rs = null;

        try {

            stmt = con.prepareStatement("SELECT * FROM PRODUTOS");
            rs = stmt.executeQuery();

            //enquanto existirem produtos no banco de dados

            while(rs.next()){
                Produto produto = new Produto();
                produto.setID(rs.getLong("PRODUTOS_ID"));
                produto.setNome(rs.getString("NOME"));
                produto.setDescricao(rs.getString("DESCRICAO"));
                produto.setMarca(rs.getString("MARCA"));
                produto.setCategoria(rs.getString("CATEGORIA"));
                produto.setPreco(rs.getDouble("PRECO"));
                produto.setCusto(rs.getDouble("CUSTO"));
                produtos.add(produto); // insere o produto na lista
            }
        } catch (SQLException e) {
            FabricaConexao.trataProblemaBancoDados(e);
        } finally {
            //fecha a conecao
            FabricaConexao.closeConnection(con, stmt, rs);
        }

        return produtos;
    }

    /**
     * Atualiza um produto no banco de dados
     *
     * @param p o produto a ter seus dados atualizados.
     */
    public void update( Produto p ){
        // cria conexao
        Connection con = FabricaConexao.getConection();
        PreparedStatement stmt = null;
        try {
            stmt = con.prepareStatement("UPDATE PRODUTOS SET NOME = ?, DESCRICAO = ?, MARCA = ?, CATEGORIA = ?, PRECO = ?, CUSTO = ? WHERE PRODUTOS_ID = ?");
            stmt.setString(1, p.getNome());
            stmt.setString(2, p.getDescricao());
            stmt.setString(3, p.getMarca());
            stmt.setString(4, p.getCategoria());
            stmt.setDouble(5, p.getPreco());
            stmt.setDouble(6, p.getCusto());
            stmt.setLong(7, p.getID()); // modifica o objeto pelo ID dele
            stmt.executeUpdate();
        } catch (SQLException e) {
            FabricaConexao.trataProblemaBancoDados(e);
        } finally {
            //fecha a conexao
            FabricaConexao.closeConnection(con, stmt);
        }
    }

    /**
     * Remove um produto no banco de dados
     *
     * @param p o produto a ser removido.
     */
    public void delete( Produto p ){
        // cria conexao
        Connection con = FabricaConexao.getConection();
        PreparedStatement stmt = null;

        try {
            stmt = con.prepareStatement("DELETE FROM PRODUTOS WHERE PRODUTOS_ID = ?");
            stmt.setLong(1, p.getID()); // deleta pelo ID do produto
            stmt.executeUpdate();

        } catch (SQLException e) {
            FabricaConexao.trataProblemaBancoDados(e);
        } finally {
            FabricaConexao.closeConnection(con, stmt);
        }
    }
}
